backend/modules/pdf_processing/progress_service.py

```
@@ -0,0 +1,307 @@
+    async def send_progress(self, session_id: str, progress_data: Dict[str, Any]):
+        """Send progress update to a specific session"""
+        if session_id in self.active_connections:
+            try:
+                websocket = self.active_connections[session_id]
+                await websocket.send_text(json.dumps(progress_data))
+                logger.debug(f"Sent progress update to {session_id}: {progress_data}")
+            except Exception as e:
+                logger.error(f"Failed to send progress to {session_id}: {e}")
+                self.disconnect(session_id)
+    
+    def update_progress(self, session_id: str, stage: str, progress: int, message: str = "", details: Dict[str, Any] = None):
```

Ruff (0.14.5)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)